packages/react-admin/src/sideEffect/saga/i18n.js

```diff
@@ -0,0 +1,33 @@
+import { all, call, put, select, takeLatest } from 'redux-saga/effects';
+import {
+    CHANGE_LOCALE,
+    USER_LOGOUT,
+    changeLocale,
+    changeLocaleSuccess,
+    changeLocaleFailure,
+} from '../../actions';
+import { getLocale } from '../../reducer';
+
+export default i18nProvider => {
+    function* loadMessages(action) {
+        const locale = action.payload;
+        try {
+            const messages = yield call(i18nProvider, locale);
+            yield put(changeLocaleSuccess(locale, messages));
+        } catch (err) {
+            yield put(changeLocaleFailure(locale));
+        }
+    }
+    function* loadInitialMessages() {
+        const initialLocale = yield select(getLocale);
+        yield put(changeLocale(initialLocale));
+    }
+
+    return function*() {
+        yield all([
+            takeLatest(CHANGE_LOCALE, loadMessages),
+            takeLatest(USER_LOGOUT, loadInitialMessages),
```

Comment:
What's the rationale of doing it on user logout?

Reply:
the messages are removed from the store using the `resettableAppReducer`, so they should be reloaded on user logout. Also the user locale can be different from the site default locale.

Reply:
That means, IMO, that the *initial state* of the i18n reducer is wrong.

Reply:
I see what you're saying, but my problem is with the how. The messages aren't known on initial render of `Admin`, not until the `i18n` saga finishes loading them. I can add loading state to `Admin`, but that complicates `Admin` more, any suggestions?